### PR TITLE
Setting SELinux Labels correctly in prep-shim

### LINKED ISSUE #1928: SELinux labels are not set correctly in the new prep-shim script
When the prep-shim script copies tar balls and MD5s into the archive directory (and maybe also when it creates a new controller directory), it need to set the selinux label correctly by doing the equivalent of `restorecon -R /path/to/dir`.

Apparently the label was inherited properly from the `fs-version-001` directory in the old shell script, but we seem to have lost that with the new python script.

That needs to be fixed to allow access to tar balls by users, using the `pbench-tarballs` URL.

### PR DESCRIPTION
**Pull request by riya-17**
Fixes #1928 

We not only need to preserve the labels but also all the metadata of the tarball.

- [x] restorecon - help to preserve all metadata

### CODE REVIEW

File: server/bin/pbench-server-prep-shim-002.py [hunk 1]
```diff
@@ -166,7 +166,7 @@ def process_tb(config, logger, receive_dir, qdir_md5, duplicates, errors):
         # First, copy the small .md5 file to the destination. That way, if
         # that operation fails it will fail quickly since the file is small.
         try:
-            shutil.copy2(tbmd5, dest)
+            shutil.copy(tbmd5, dest)
```
Comment (ndokos) [inline server/bin/pbench-server-prep-shim-002.py:169]: Does the `shutil.move()` call on l.190 do the right thing for the tarball? IIUC, it does not and we need to do something there. I wonder if the thing to do is leave the `shutil.copy2()` alone (again IIUC, that's the equivalent of `cp -a` in the shell) and do explicit `restorecon` calls (using the `selinux` module from the `libselinux-python3` RPM package) to set the SELinux context on the tar ball and md5 files.

**Review by ndokos — CHANGES_REQUESTED** [review]

File: server/bin/pbench-server-prep-shim-002.py (hunk 1 above)
Comment (riya-17) [reply server/bin/pbench-server-prep-shim-002.py]: seems fair

Comment (riya-17) [reply server/bin/pbench-server-prep-shim-002.py]: How should I import SELinux in code? when I try the python3 command line, it is working after importing. In code, it is saying no module named "selinux".

Comment (riya-17) [reply server/bin/pbench-server-prep-shim-002.py]: Do we need to define in PATH?

Comment (ndokos) [reply server/bin/pbench-server-prep-shim-002.py]: If you have the `libeslinux-python3` RPM installed, you should have an `selinux` python3 module in `/usr/lib64/python3.6/site-packages/selinux` (on RHEL7). Then you should be able to
```
import selinux
...
selinux.restorecon(<path>)
...
```
Does that not work?

**Review by ndokos — COMMENTED** [review]
I think that *should* do it. Do we have unit tests for checking? Do the current unit tests pass locally?

You will need to rebase too.

**Comment by riya-17** [comment]
> I think that _should_ do it. Do we have unit tests for checking? Do the current unit tests pass locally?
> 
> You will need to rebase too.

No, I tested locally but we do not have unittests.  Yaa current unit tests are passing locally.
Yeah I will rebase it

File: server/bin/pbench-server-prep-shim-002.py (hunk 1 above)
Comment (riya-17) [reply server/bin/pbench-server-prep-shim-002.py]: I am using RHEL8 and It is failing test cases.

**Comment by riya-17** [comment]
Facing this error in RHEL8
`! Traceback (most recent call last):
!   File "/var/tmp/pbench-test-server/test-0.1/opt/pbench-server/bin/pbench-server-prep-shim-002.py", line 15, in <module>
!     import selinux
! ModuleNotFoundError: No module named 'selinux'`

File: server/bin/pbench-server-prep-shim-002.py [hunk 2]
```diff
@@ -167,6 +168,7 @@ def process_tb(config, logger, receive_dir, qdir_md5, duplicates, errors):
         # that operation fails it will fail quickly since the file is small.
         try:
             shutil.copy2(tbmd5, dest)
+            selinux.restorecon(dest / tbmd5.name)
```
Comment (dbutenhof) [inline server/bin/pbench-server-prep-shim-002.py:171]: According to my reading, this basically sets default context for the target file. Is it ever possible that the original file in the reception area might have any non-default context? If so, we're throwing it away here.

Comment (riya-17) [reply server/bin/pbench-server-prep-shim-002.py:171]: No idea about that. @ndokos

Comment (portante) [reply server/bin/pbench-server-prep-shim-002.py:171]: If you install this on a VM can we try out a few scenarios to see if this works properly?  Can you devise a test to expose this behavior?

**Review by portante — CHANGES_REQUESTED** [review]

**Comment by dbutenhof** [comment]
It's unfortunate that you apparently couldn't get the selinux Python module to work for you. That's a mystery... and it seems a shame to revert to just spawning the command because that's a lot less efficient. (Presuming of course that the Python module wasn't already doing that!)

In any case, we definitely need to try to confirm proper behavior with a unit test.

**Review by ndokos — CHANGES_REQUESTED** [review]
Why is the module insufficient? There is no version of it for the SCL environment AFAICT, but there are packages for the native python3 environment:

- F31 (and I presume F32 and F33): `python3-libselinux.x86_64                                                                         2.9-6.fc31                                                                          @updates`
- RHEL7: `libselinux-python3.x86_64             2.5-15.el7             @rhel-7-server-rpms`
- RHEL8: `python3-libselinux.x86_64       2.9-3.el8          @anaconda` (plus there is an update available: `python3-libselinux.x86_64       2.9-4.el8_3        rhel-8-for-x86_64-baseos-rpms`

So apart from the screwy naming on RHEL7, there should be no problem in using the `selinux` python module in all platforms. If there are problems, let's discuss in the scrum.

**Review by dbutenhof — CHANGES_REQUESTED** [review]
Actually, the containerized Jenkins tests are failing because `selinux` isn't installed. I don't think it was already in the `requirements.txt` file, so you probably need to add it.

**Review by portante — CHANGES_REQUESTED** [review]
Seems fine, but needs a rebase.

@riya-17, how does the test output compare with `test-5.2.txt` on the `b0.69` branch?

File: server/requirements.txt [hunk 3]
```diff
@@ -7,3 +7,4 @@ gunicorn
 humanize
 pyesbulk>=2.0.1
 requests
+libselinux
```
Comment (dbutenhof) [inline server/requirements.txt:10]: I think this should be `selinux`; at least, `pip` sees that module, and it can be imported, whereas attempting to `pip install libselinux` locally gives me the same problem as the Jenkins tests.

**Review by dbutenhof — CHANGES_REQUESTED** [review]
```
[31mERROR: Could not find a version that satisfies the requirement libselinux (from -r /home/pbench/server/requirements.txt (line 10)) (from versions: none)[0m
[31mERROR: No matching distribution found for libselinux (from -r /home/pbench/server/requirements.txt (line 10))[0m
```

**Comment by portante** [comment]
Let's get this rebased first and address the software dependency requirements before we pull it back out of `Draft` mode.

**Comment by riya-17** [comment]
> Why is the module insufficient? There is no version of it for the SCL environment AFAICT, but there are packages for the native python3 environment:
> 
>     * F31 (and I presume F32 and F33): `python3-libselinux.x86_64                                                                         2.9-6.fc31                                                                          @updates`
> 
>     * RHEL7: `libselinux-python3.x86_64             2.5-15.el7             @rhel-7-server-rpms`
> 
>     * RHEL8: `python3-libselinux.x86_64       2.9-3.el8          @anaconda` (plus there is an update available: `python3-libselinux.x86_64       2.9-4.el8_3        rhel-8-for-x86_64-baseos-rpms`
> 
> 
> So apart from the screwy naming on RHEL7, there should be no problem in using the `selinux` python module in all platforms. If there are problems, let's discuss in the scrum.

These are the software dependency requirements as mentioned by @ndokos and these are only available with dnf install

**Review by ndokos — CHANGES_REQUESTED** [review]
IIUC, we need to test this on a RHEL7 VM *and* a RHEL8 VM to make sure that the tests pass with the given dependencies. And we also need to add some tests for the specific behaviors we are looking for (in particular, that a tar ball that is copied into the archive has the `httpd_<mumble>` label. Is there anything else to do here?

File: server/test-requirements.txt [hunk 4]
```diff
@@ -6,3 +6,4 @@ pytest>=4.6.3, < 5
 pytest-helpers-namespace>=2019.1.8
 pytest-mock>=1.10.4
 requests_mock
+selinux
```
Comment (portante) [inline server/test-requirements.txt:9]: You don't need this in `test-requirements.txt` if you already have it in `requirements.txt`.

File: server/bin/gold/test-5.2.txt [hunk 5]
```diff
@@ -1,4 +1,12 @@
...
+Warning no default label for /var/tmp/pbench-test-server/test-5.2/pbench/archive/fs-version-001/controller-b-with-prefixes/tarball-0_1970.01.01T00.42.00.tar.xz.md5
+Warning no default label for /var/tmp/pbench-test-server/test-5.2/pbench/archive/fs-version-001/controller-b-with-prefixes/tarball-w-dot-prefix_1970.01.01T00.42.00.tar.xz
+Warning no default label for /var/tmp/pbench-test-server/test-5.2/pbench/archive/fs-version-001/controller-b-with-prefixes/tarball-w-dot-prefix_1970.01.01T00.42.00.tar.xz.md5
+Warning no default label for /var/tmp/pbench-test-server/test-5.2/pbench/archive/fs-version-001/controller-b-with-prefixes/tarball-w-prefix-dot_1970.01.01T00.42.00.tar.xz
+Warning no default label for /var/tmp/pbench-test-server/test-5.2/pbench/archive/fs-version-001/controller-b-with-prefixes/tarball-w-prefix-dot_1970.01.01T00.42.00.tar.xz.md5
+Warning no default label for /var/tmp/pbench-test-server/test-5.2/pbench/archive/fs-version-001/controller-g-normal/tarball-normal_1970.01.01T00.42.00.tar.xz
+Warning no default label for /var/tmp/pbench-test-server/test-5.2/pbench/archive/fs-version-001/controller-g-normal/tarball-normal_1970.01.01T00.42.00.tar.xz.md5
```
Comment (portante) [inline server/bin/gold/test-5.2.txt:9]: Who generates this output?  If this is coming from the selinux module itself, we'll have to get a hold of how it is produced.

**Review by portante — CHANGES_REQUESTED** [review]

File: server/test-requirements.txt (hunk 4 above)
Comment (riya-17) [reply server/test-requirements.txt:9]: oh ok... I will change that

File: server/bin/gold/test-5.2.txt (hunk 5 above)
Comment (riya-17) [reply server/bin/gold/test-5.2.txt:9]: yeah It is coming from the SELinux module I will see how it is getting produced.

File: server/requirements.txt (hunk 3 above)
Comment (riya-17) [reply server/requirements.txt]: yes, I changed that

File: server/bin/gold/test-5.2.txt (hunk 5 above)
Comment (riya-17) [reply server/bin/gold/test-5.2.txt:9]: "libselinux: selinux_restorecon: only log no default label warning for caller-supplied pathname", I will look how we can disable this warning

**Review by portante — CHANGES_REQUESTED** [review]
Okay this looks good, but the unit testing needs to be addressed still.

In particular, we need to "mock out" the actual SELinux operations, since for unit tests we only need to verify that we make the proper API call with the correct arguments.

I'd suggest that we `import` a mocked out module of the SELinux module in `pbench-server-prep-shim-002.py` when running under the unit tests, where that mock to records the argument pass to `selinux.restorecon()`

However, we also need to check that we have a proper label for the argument.  So perhaps we can wrap the `selinux.restorecon()` call to capture `stdout` and `stderr` and report our own failure if something was captured?

**Comment by riya-17** [comment]
> Okay this looks good, but the unit testing needs to be addressed still.
> 
> In particular, we need to "mock out" the actual SELinux operations, since for unit tests we only need to verify that we make the proper API call with the correct arguments.
> 
> I'd suggest that we `import` a mocked out module of the SELinux module in `pbench-server-prep-shim-002.py` when running under the unit tests, where that mock to records the argument pass to `selinux.restorecon()`
> 
> However, we also need to check that we have a proper label for the argument. So perhaps we can wrap the `selinux.restorecon()` call to capture `stdout` and `stderr` and report our own failure if something was captured?

yeah, unit testing is giving us a hard time so working on that. I will create a unit test for verifying the behavior.

File: lib/pbench/server/selinux.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    force -- Force reset of context to match file_context for customizable files,
+    and the default file context, changing the user, role, range portion  as well
+    as the type (default False)
+    """
+    restorecon_flags = False
+    status = selinux_restorecon(os.path.expanduser(path), restorecon_flags)
+    return status
+
+
+def selinux_restorecon(pathname, restorecon_flags):
```
Comment (portante) [inline lib/pbench/server/selinux.py:23]: When not running the unit tests, how does this work?

Comment (riya-17) [reply lib/pbench/server/selinux.py:23]: yeah, that's what I am working on.. just pushed the work.

Comment (riya-17) [reply lib/pbench/server/selinux.py:23]: just wanted to see whether it is working on Jenkins majorly

**Review by ndokos — DISMISSED** [review]

**Comment by ndokos** [comment]
Needs rebasing.

File: server/bin/test-bin/selinux.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    Arguments:
+    path -- The pathname for the file or directory to be relabeled.
+
+    Keyword arguments:
+    recursive -- Change files and directories file labels recursively (default False)
+    verbose -- Show changes in file labels (default False)
+    force -- Force reset of context to match file_context for customizable files,
+    and the default file context, changing the user, role, range portion  as well
+    as the type (default False)
+    """
```
Comment (portante) [inline server/bin/test-bin/selinux.py:16]: Much like we do in https://github.com/distributed-system-analysis/pbench/blob/main/agent/util-scripts/test-bin/mpstat, can you write the argument names and values to the test-execution file?

Something like:

```
import sys
import os

with open(os.environ["_testlog"], "a") as ofp:
    ofp.write(f"{path!r}, recursive={recursive!r}, verbose={verbose!r}, force={force!r}\n")

sys.exit(0)
```

File: server/bin/test-bin/selinux.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# This is a mock file for selinux.restorecon test cases
```
Comment (portante) [inline server/bin/test-bin/selinux.py:1]: Great idea for this mock module.  Can we create a new directory, `test-lib`, as a sister directory to `test-bin`, moving this module there, and adding it to the `PYTHONPATH` from there in `unittests`?  I don't think it needs to be installed in the test environment, but it probably should have an empty `__init__.py` to be a proper python3 module.

**Review by portante — COMMENTED** [review]
Looks good, just need execution output added, and the module moved.

File: server/bin/test-bin/selinux.py (hunk 7 above)
Comment (riya-17) [reply server/bin/test-bin/selinux.py:16]: ok

File: server/bin/test-bin/selinux.py (hunk 8 above)
Comment (riya-17) [reply server/bin/test-bin/selinux.py:1]: Ok sure will do that

File: server/bin/unittests [hunk 9]
```diff
@@ -250,6 +250,10 @@ function _setup_state {
     let res=res+$?
     cp $_tdir/test-bin/* $_testopt_sat/unittest-scripts/
     let res=res+$?
+    cp -a $_tdir/test-lib/* $_testopt/unittest-scripts/
```
Comment (portante) [inline server/bin/unittests:253]: No sure this should be in `unittest-scripts`; can we consider something like:

```
mkdir ${_testopt}/unittest-lib
cp -a ${_tdir}/test-lib/* ${_testopt}/unittest-lib/
...
```

**Review by portante — CHANGES_REQUESTED** [review]

File: server/bin/unittests (hunk 9 above)
Comment (riya-17) [reply server/bin/unittests]: yeah sure

**Review by ndokos — DISMISSED** [review]

File: server/bin/pbench-server-prep-shim-002.py (hunk 1 above)
Comment (portante) [reply server/bin/pbench-server-prep-shim-002.py]: I believe this comment is now resolved.

File: server/test-requirements.txt (hunk 4 above)
Comment (portante) [reply server/test-requirements.txt]: I see that we have resolved this by using an RPM in the development container.

We also need to then add this in the pbench-server RPM `.spec` file.

But, we should also have a python3 library module requirement in the `server/requirements.txt` file.  Can that be restored?

File: server/bin/gold/test-5.2.txt (hunk 5 above)
Comment (portante) [reply server/bin/gold/test-5.2.txt]: I believe this has been resolved.

File: lib/pbench/server/selinux.py (hunk 6 above)
Comment (portante) [reply lib/pbench/server/selinux.py]: I believe this has been resolved.

**Review by portante — CHANGES_REQUESTED** [review]
This looks really good.  Just one more topic to address:

We need to add the required RPM to the pbench-server RPM .spec file.

And, we should also have a python3 library module requirement in the server/requirements.txt file.

**Comment by riya-17** [comment]
> This looks really good. Just one more topic to address:
> 
> We need to add the required RPM to the pbench-server RPM .spec file.
> 
> And, we should also have a python3 library module requirement in the server/requirements.txt file.

Hey, are you talking about the spec file at location pbench/server/rpm/pbench-server.spec.j2?

adding the python3 library module in requirements.txt file might create problem coz it will then consider the pip installed module rather than rpm module which our code should not prefer as there are no restorecons available in that

**Comment by portante** [comment]
> Hey, are you talking about the spec file at location pbench/server/rpm/pbench-server.spec.j2?

Yes.

> Adding the python3 library module in the `requirements.txt` file might create a problem coz it will then consider the pip installed module rather than the rpm module which our code should not prefer as there are no properly build SELinux modules available with that especially for fedora 32.

Ah, right, this is the problem that the RPM provides the Python3 module, but it is not available as a pip3 installable module as well.

I'd add a comment to the `server/requirements.txt` file to state that we don't have a pip3 module for selinux, but that it is provided by the selinux RPM for the various distributions we support.

File: server/rpm/pbench-server.spec.j2 [hunk 10]
```diff
@@ -65,6 +65,9 @@ cd /%{installdir}
 rm -rf node_modules
 npm install
 
+# install selinux modules under /%{installdir}
+dnf install python3-libselinux
```
Comment (portante) [inline server/rpm/pbench-server.spec.j2:69]: Actually, since we are in an RPM `.spec` file, we can add a `Requires: python3-libselinux` line to replace line 21.

**Review by portante — CHANGES_REQUESTED** [review]

File: server/rpm/pbench-server.spec.j2 (hunk 10 above)
Comment (riya-17) [reply server/rpm/pbench-server.spec.j2:69]: ohh ok.. sorry I was not sure how I can add

**Review by portante — APPROVED** [review]

**Review by ndokos — DISMISSED** [review]

File: server/bin/unittests [hunk 11]
```diff
@@ -315,6 +323,7 @@ function _setup_state {
     # are found in $_testopt/unittest-scripts.
     _orig_PATH=$PATH
     export PATH=$_testopt/unittest-scripts:$_testopt/bin:$PATH
+    export PYTHONPATH=$_testopt/unittest-lib:$PYTHONPATH
```
Comment (dbutenhof) [inline server/bin/unittests:326]: You copy into separate directories for "server" and "satellite", but there's a single `PYTHONPATH` modification. Is that really what we want? Is there any need for separate Python overlay libraries? Is doesn't appear there is right now, so maybe reserve that for later and just make the single copy you're using?

**Review by dbutenhof — APPROVED** [review]

**Review by npalaska — DISMISSED** [review]

File: server/bin/unittests (hunk 11 above)
Comment (portante) [reply server/bin/unittests:326]: At the risk of belaboring this patch, even though it does still need a rebase, this seems like a good catch that we should address.

We handle the `PATH` for the satellite script invocation context separately at: https://github.com/distributed-system-analysis/pbench/blob/3219eea0d1fce179cd1d71ef3b6a1227ec57c842/server/bin/unittests#L111

So we should similarly setup `PYTHONPATH` to add the satellite test libs there as well.

**Review by portante — CHANGES_REQUESTED** [review]
... and a rebase ...

**Comment by riya-17** [comment]
> ... and a rebase ...

Done and fixed

**Review by dbutenhof — APPROVED** [review]

File: server/bin/unittests [hunk 12]
```diff
@@ -109,6 +109,7 @@ function _run_allscripts {
     _run pbench-verify-backup-tarballs
     # Invoke pbench-satellite-cleanup in the context of the satellite
     PATH=${_testopt_sat}/unittest-scripts:${_testopt_sat}/bin:${_orig_PATH} _PBENCH_SERVER_CONFIG=${SATCONFIG} _run pbench-satellite-cleanup
+    PYTHONPATH=${_testopt_sat}/unittest-lib:${_orig_PYTHONPATH}
```
Comment (portante) [inline server/bin/unittests:112]: This is subtle:  the `PATH` definition on line 111 above is actually using `bash`'s "temporary environment variable export" feature, where the updated `PATH` assignment is only enabled for the duration of the `_run` invocation on line 111.

In order for this to work, we want to move this `PYTHONPATH` setting following that `PATH`.

Try this out in `bash` to see how the two variables used by `ls` are only set for the duration of the command:

```
$ unset LS_COLORS
$ unset TIME_STYLE
$ ls -l
$ TIME_STYLE=full-iso LS_COLORS=always ls -l
$ echo "TIME_STYLE='${TIME_STYLE}'; LS_COLORS='${LS_COLORS}'"
$ dircolors    # restores your terminal ls colors
```

**Review by portante — CHANGES_REQUESTED** [review]

File: server/bin/unittests (hunk 12 above)
Comment (dbutenhof) [reply server/bin/unittests:112]: Good catch; the `PATH` assignment was long enough that I missed that there was a command at the end of the line.

So this `PYTHONPATH` needs to be moved up to 111 along with `PATH` to have effect on `_run`.

Comment (ndokos) [reply server/bin/unittests:112]: Maybe break it into three (or more) lines with `\` to escape the unwanted newlines?

```
PATH=...... \
PYTHONPATH=.... \
_PBENCH_SERVER_CONFIG=.... \
command
```

**Review by ndokos — COMMENTED** [review]
Just a comment on @portante's change request.

File: server/bin/unittests (hunk 12 above)
Comment (riya-17) [reply server/bin/unittests]: I didn't know that. Thank you.  Fixed!

**Review by portante — APPROVED** [review]

**Review by ndokos — APPROVED** [review]

**Review by dbutenhof — APPROVED** [review]

**Review by portante — APPROVED** [review]
